pkg/faketikv/node.go

```diff
@@ -0,0 +1,213 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	store := &metapb.Store{
+		Id:      id,
+		Address: addr,
+	}
+	stats := &pdpb.StoreStats{
+		StoreId: id,
+		// TODO: configurable
+		Capacity:  1000000000000,
+		Available: 1000000000000,
+		StartTime: uint32(time.Now().Unix()),
+	}
+	client, reciveRegionHeartbeatCh, err := NewClient(pdAddr)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	return &Node{
+		Store:  store,
+		stats:  stats,
+		client: client,
+		ctx:    ctx,
+		cancel: cancel,
+		tasks:  make(map[uint64]Task),
+		state:  Down,
+		reciveRegionHeartbeatCh: reciveRegionHeartbeatCh,
+	}, nil
+}
+
+// Start starts the node.
+func (n *Node) Start() error {
+	ctx, cancel := context.WithTimeout(n.ctx, pdTimeout)
+	err := n.client.PutStore(ctx, n.Store)
+	cancel()
+	if err != nil {
+		return err
+	}
+	n.wg.Add(1)
+	go n.reciveRegionHeartbeat()
+	n.state = Up
+	return nil
+}
+
+func (n *Node) reciveRegionHeartbeat() {
+	defer n.wg.Done()
+	for {
+		select {
+		case resp := <-n.reciveRegionHeartbeatCh:
+			task := responseToTask(resp, n.clusterInfo)
+			if task != nil {
+				n.clusterInfo.AddTask(task)
+			}
+		case <-n.ctx.Done():
+			return
+		}
+	}
+}
+
+// Tick steps node status change.
+func (n *Node) Tick() {
+	if n.state != Up {
+		return
+	}
+	n.stepHeartBeat()
+	n.stepTask()
+	// step regions will step region status. like elect a leader, split region and so on.
+	n.clusterInfo.stepRegions()
+	n.tick++
+}
+
+// GetState returns current node state.
+func (n *Node) GetState() NodeState {
+	return n.state
+}
+
+func (n *Node) stepTask() {
+	n.Lock()
+	defer n.Unlock()
+	for _, task := range n.tasks {
+		task.Step(n.clusterInfo)
+		if task.IsFinished() {
+			log.Infof("[store %d][region %d] task finished: %s", n.GetId(), task.RegionID(), task.Desc())
+			n.clusterInfo.reportRegionChange(task.RegionID())
+			delete(n.tasks, task.RegionID())
+		}
+	}
+}
+
+func (n *Node) stepHeartBeat() {
+	if n.tick%storeHeartBeatPeriod == 0 {
+		n.storeHeartBeat()
+	}
+	if n.tick%regionHeartBeatPeriod == 0 {
+		n.regionHeartBeat()
+	}
+}
+
+func (n *Node) storeHeartBeat() {
+	if n.state != Up {
+		return
+	}
+	ctx, cancel := context.WithTimeout(n.ctx, pdTimeout)
+	err := n.client.StoreHeartbeat(ctx, n.stats)
+	if err != nil {
+		log.Infof("[store %d] report heartbeat error: %s", n.GetId(), err)
+	}
+	cancel()
+}
+
+func (n *Node) regionHeartBeat() {
+	regions := n.clusterInfo.GetRegions()
+	for _, region := range regions {
+		if region.Leader != nil && region.Leader.GetStoreId() == n.Id {
+			ctx, cancel := context.WithTimeout(n.ctx, pdTimeout)
+			err := n.client.RegionHeartbeat(ctx, region)
+			if err != nil {
+				log.Infof("[node %d][region %d] report heartbeat error: %s", n.Id, region.GetId(), err)
+			}
+			cancel()
+		}
+	}
+}
+
+func (n *Node) reportRegionChange(regionID uint64) {
+	region := n.clusterInfo.GetRegion(regionID)
+	if region.Leader.GetStoreId() == n.Id {
```

Comment:
why check store id here? https://github.com/nolouch/pd/blob/79ccfc080b6de14f5d3bf6c5dd32d2ff02541c4d/pkg/faketikv/cluster.go#L85   seem that it can make sure region.Leader.GetStoreId == n.Id

Reply:
should make sure node report it's leader heartbeat no matter other places call it.